specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,501 @@
+
+### CrossDomainMessenger
+
+The `CrossDomainMessenger` contracts on both L1 and L2 MUST NOT be able to faciliate the transfer of native asset
+on custom gas token chains due to their tight coupling with `ether` and the `OptimismPortal` and `L2ToL1MessagePasser`
+contracts.
+
+It is possible to support this in the future with a redesign of the `CrossDomainMessenger` contract.
+It would need to have conditional logic based on being a custom gas token chain and faciliate transfer
+of the ERC20 token to the end user on the other side. It adds a layer of extra state modifying calls
+so it may not be worth adding.
+
+The following methods MUST revert when `CALLVALUE` is non zero:
+
+- `sendMessage(address,bytes,uint32)`
+
+It should be impossible to call `relayMessage` when `CALLVALUE` is non zero, as it is enforced by `sendMessage`.
+
+The `CrossDomainMessenger` also has the API for _getting_ the custom gas token, namely `gasPayingToken()`, which outputs
+a tuple of the address and decimals of the custom gas token.
+
+- The `L1CrossDomainMessenger` fetches this tuple from the `SystemConfig` contract.
+- The `L2CrossDomainMessenger` fetches this tuple from the `L1Block` contract.
+
+### SystemConfig
+
+The `SystemConfig` is the source of truth for the address of the custom gas token. It does on chain validation,
+stores information about the token and well as passes the information to L2.
+
+#### initialize
+
+The `SystemConfig` is modified to allow the address of the custom gas paying token to be set during the call
+to `initialize`. Using a custom gas token is indicated by passing an address other than `ETHER_TOKEN_ADDRESS`
+or `address(0)`. If `address(0)` is used for initialization, it MUST be mapped into `ETHER_TOKEN_ADDRESS`
+before being forwarded to the rest of the system. When a custom gas token is set, the number of decimals
+on the token MUST be exactly 18, the name of the token MUST be less than or equal to 32 bytes and the
+symbol MUST be less than or equal to 32 bytes. If the token passes all of these checks,
+`OptimismPortal.setGasPayingToken` is called. The implementation of `initialize` MUST not allow the chain
+operator to change the address of the custom gas token if it is already set.
+
+### L1Block
+
+The `L1Block` contract is upgraded with a permissioned function `setGasPayingToken` that is used to set
+information about the gas paying token. The `DEPOSITOR_ACCOUNT` MUST be the only account that can call the
+setter function. This setter is differentiated from the `setL1BlockValues` functions because the gas paying
+token is not meant to be dynamic config whereas the arguments to `setL1BlockValues` are generally dynamic in nature.
+
+Any L2 contract that wants to learn the address of the gas paying token can call the getter on the
+`L1Block` contract.
+
+```solidity
+function setGasPayingToken(address _token, uint8 _decimals, byte32 _name, bytes32 _symbol) external;
+```
```

Comment:
Can we specify the accessor here as well and move the mutator above the mention of the getter?
```suggestion
```solidity
function setGasPayingToken(address _token, uint8 _decimals, byte32 _name, bytes32 _symbol) external;
```

Any L2 contract that wants to learn the address of the gas paying token can call the getter on the
`L1Block` contract.

```solidity
function gasPayingToken() external view returns (address, uint8);
```
```

Reply:
This is covered in the `IGasToken` section so i dont think this is necessary